File: sdks/typescript/src/stream.ts
/**
 * Synap TypeScript SDK - Event Stream Module
 * 
 * Event stream operations with reactive consumption patterns.
 */

import { Observable, Subject, timer, EMPTY, defer } from 'rxjs';
import { 
  switchMap, 
  retry, 
  catchError, 
  filter, 
  takeUntil,
  share,
  map
} from 'rxjs/operators';
import type { SynapClient } from './client';
import type {
  StreamEvent,
  StreamPublishOptions,
  StreamStats,
  StreamConsumerOptions,
  ProcessedStreamEvent,
} from './types';

/**
 * Event Stream Manager with reactive support
 */
export class StreamManager {
  private stopSignals = new Map<string, Subject<void>>();

  constructor(private client: SynapClient) {}

  /**
   * Create a new stream room
   */
  async createRoom(roomName: string): Promise<boolean> {
    const result = await this.client.sendCommand<{ success: boolean }>('stream.create', {
      room: roomName,
    });
    return result.success;
  }

  /**
   * Publish an event to a stream room
   */
  async publish(
    roomName: string,
    eventName: string,
    data: any,
    options?: StreamPublishOptions
  ): Promise<number> {
    const payload: Record<string, any> = {
      room: roomName,
      event: eventName,
      data,
    };

    if (options?.metadata) {
      payload.metadata = options.metadata;
    }

    const result = await this.client.sendCommand<{ offset: number }>(
      'stream.publish',
      payload
    );

    return result.offset;
  }

  /**
   * Consume events from a stream room (one-time fetch)
   */
  async consume(
    roomName: string,
    subscriberId: string,
    fromOffset: number = 0
  ): Promise<StreamEvent[]> {
    const result = await this.client.sendCommand<{ events: any[] }>(
      'stream.consume',
      {
        room: roomName,
        subscriber_id: subscriberId,
        from_offset: fromOffset,
      }
    );

    const events = result.events || [];
    
    // Parse events - convert byte array data to objects
    return events.map(event => ({
      ...event,
      data: this.parseEventData(event.data),
    }));
  }

  /**
   * Parse event data - handle both byte arrays and objects
   */
  private parseEventData(data: any): any {
    if (Array.isArray(data)) {
      // Convert byte array to string and parse as JSON
      try {
        const text = new TextDecoder().decode(new Uint8Array(data));
        return JSON.parse(text);
      } catch (error) {
        console.error('Failed to parse event data:', error);
        return data;
      }
    }
    return data;
  }

  /**
   * Get stream room statistics
   */
  async stats(roomName: string): Promise<StreamStats> {
    return this.client.sendCommand<StreamStats>('stream.stats', {
      room: roomName,
    });
  }

  /**
   * List all stream rooms
   */
  async listRooms(): Promise<string[]> {
    const result = await this.client.sendCommand<{ rooms: string[] }>('stream.list', {});
    return result.rooms;
  }

  /**
   * Delete a stream room
   */
  async deleteRoom(roomName: string): Promise<boolean> {
    const result = await this.client.sendCommand<{ deleted: any }>('stream.delete', {
      room: roomName,
    });
    // Server returns room name as confirmation, treat as success
    return !!result.deleted;
  }

  // ==================== REACTIVE METHODS ====================

  /**
   * Create a reactive event consumer as an Observable
   * 
   * @example
   * ```typescript
   * synap.stream.consume$({
   *   roomName: 'chat-room',
   *   subscriberId: 'user-123',
   *   fromOffset: 0,
   *   pollingInterval: 500
   * }).subscribe({
   *   next: (event) => {
   *     console.log('Event:', event.event, event.data);
   *   }
   * });
   * ```
   */
  consume$<T = any>(options: StreamConsumerOptions): Observable<ProcessedStreamEvent<T>> {
    const {
      roomName,
      subscriberId,
      fromOffset = 0,
      pollingInterval = 1000,
    } = options;

    const stopKey = `${roomName}:${subscriberId}`;
    const stopSignal = new Subject<void>();
    this.stopSignals.set(stopKey, stopSignal);

    let currentOffset = fromOffset;

    return timer(0, pollingInterval).pipe(
      takeUntil(stopSignal),
      switchMap(async () => {
        try {
          const events = await this.consume(roomName, subscriberId, currentOffset);
          
          if (events.length > 0) {
            // Update offset to last event + 1
            currentOffset = events[events.length - 1].offset + 1;
          }
          
          return events;
        } catch (error) {
          console.error(`Error consuming from stream ${roomName}:`, error);
          return [];
        }
      }),
      // Flatten array of events into individual emissions
      switchMap((events) => events),
      map((event): ProcessedStreamEvent<T> => ({
        offset: event.offset,
        event: event.event,
        data: event.data as T,
        timestamp: event.timestamp,
      })),
      share()
    ) as Observable<ProcessedStreamEvent<T>>;
  }

  /**
   * Create a reactive consumer that filters by event name
   * 
   * @example
   * ```typescript
   * synap.stream.consumeEvent$({
   *   roomName: 'notifications',
   *   subscriberId: 'user-1',
   *   eventName: 'user.created'
   * }).subscribe({
   *   next: (event) => console.log('User created:', event.data)
   * });
   * ```
   */
  consumeEvent$<T = any>(
    options: StreamConsumerOptions & { eventName: string }
  ): Observable<ProcessedStreamEvent<T>> {
    const { eventName, ...consumeOptions } = options;
    
    return this.consume$<T>(consumeOptions).pipe(
      filter((event) => event.event === eventName)
    );
  }

  /**
   * Stop a reactive consumer
   * 
   * @param roomName - Room name
   * @param subscriberId - Subscriber ID
   */
  stopConsumer(roomName: string, subscriberId: string): void {
    const stopKey = `${roomName}:${subscriberId}`;
    const stopSignal = this.stopSignals.get(stopKey);
    
    if (stopSignal) {
      stopSignal.next();
      stopSignal.complete();
      this.stopSignals.delete(stopKey);
    }
  }

  /**
   * Stop all reactive consumers
   */
  stopAllConsumers(): void {
    this.stopSignals.forEach((signal) => {
      signal.next();
      signal.complete();
    });
    this.stopSignals.clear();
  }

  /**
   * Create an observable that emits stream statistics at regular intervals
   * 
   * @param roomName - Room name
   * @param interval - Polling interval in milliseconds (default: 5000)
   * 
   * @example
   * ```typescript
   * synap.stream.stats$('chat-room', 3000).subscribe({
   *   next: (stats) => console.log('Events:', stats.event_count),
   * });
   * ```
   */
  stats$(roomName: string, interval: number = 5000): Observable<StreamStats> {
    return timer(0, interval).pipe(
      switchMap(() => defer(() => this.stats(roomName))),
      retry({
        count: 3,
        delay: 1000,
      }),
      catchError((error: unknown) => {
        console.error(`Error fetching stats for ${roomName}:`, error);
        return EMPTY;
      }),
      share()
    );
  }

  /**
   * Helper: Publish a typed event
   */
  async publishEvent<T>(
    roomName: string,
    eventName: string,
    data: T,
    options?: StreamPublishOptions
  ): Promise<number> {
    return this.publish(roomName, eventName, data, options);
  }
}

File: sdks/rust/src/stream.rs
//! Event Stream operations

use crate::client::SynapClient;
use crate::error::Result;
use crate::types::{Event, StreamStats};
use serde::Deserialize;
use serde_json::{Value, json};

/// Wire format of a stream event as returned by the server.
/// HTTP returns `data` as `Vec<u8>` (serde_json::to_vec of the original JSON).
/// SynapRPC may return `data` as a string.  We accept both via `Value`.
#[derive(Deserialize)]
struct RawStreamEvent {
    #[serde(default)]
    offset: u64,
    event: String,
    data: Value,
    #[serde(default)]
    timestamp: Option<u64>,
}

impl From<RawStreamEvent> for Event {
    fn from(raw: RawStreamEvent) -> Self {
        let data = match &raw.data {
            // HTTP: data arrives as JSON array of bytes → decode back to JSON
            Value::Array(arr) => {
                let bytes: Vec<u8> = arr
                    .iter()
                    .filter_map(|v| v.as_u64().map(|n| n as u8))
                    .collect();
                serde_json::from_slice(&bytes).unwrap_or(Value::Null)
            }
            // RPC: data arrives as a JSON string (the serialized JSON)
            Value::String(s) => serde_json::from_str(s).unwrap_or(Value::String(s.clone())),
            // Already a JSON value (unlikely but handle gracefully)
            other => other.clone(),
        };
        Self {
            offset: raw.offset,
            event: raw.event,
            data,
            timestamp: raw.timestamp,
        }
    }
}

/// Stream Manager interface
///
/// Uses StreamableHTTP protocol for all operations.
/// Event Streams are **reactive by default** - use `observe_events()` or `observe_event()`.
#[derive(Clone)]
pub struct StreamManager {
    pub(crate) client: SynapClient,
}

impl StreamManager {
    /// Create a new Stream manager interface
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Create a new stream room
    pub async fn create_room(&self, room: &str, max_events: Option<usize>) -> Result<()> {
        let mut payload = json!({"room": room});
        if let Some(max) = max_events {
            payload["max_events"] = json!(max);
        }

        self.client.send_command("stream.create", payload).await?;
        Ok(())
    }

    /// Publish an event to a stream
    ///
    /// # Returns
    /// Returns the offset of the published event
    ///
    /// # Example
    /// ```no_run
    /// # use synap_sdk::{SynapClient, SynapConfig};
    /// # use serde_json::json;
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// # let client = SynapClient::new(SynapConfig::new("http://localhost:15500"))?;
    /// let offset = client.stream().publish(
    ///     "chat-room",
    ///     "message",
    ///     json!({"user": "alice", "text": "Hello!"})
    /// ).await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn publish(&self, room: &str, event: &str, data: Value) -> Result<u64> {
        let payload = json!({
            "room": room,
            "event": event,
            "data": data,
        });

        let response = self.client.send_command("stream.publish", payload).await?;

        Ok(response["offset"].as_u64().unwrap_or(0))
    }

    /// Consume events from a stream
    ///
    /// # Arguments
    /// * `room` - Stream room name
    /// * `offset` - Starting offset (None = from beginning)
    /// * `limit` - Maximum events to fetch
    pub async fn consume(
        &self,
        room: &str,
        offset: Option<u64>,
        limit: Option<usize>,
    ) -> Result<Vec<Event>> {
        let payload = json!({
            "room": room,
            "subscriber_id": "sdk-default",
            "from_offset": offset.unwrap_or(0),
            "limit": limit,
        });

        let response = self.client.send_command("stream.consume", payload).await?;

        // Server stores data as Vec<u8> (serde_json::to_vec) and serialises the
        // struct directly, so `data` arrives as a JSON byte-array.  We decode
        // each event's data bytes back into the original JSON value.
        let raw_events: Vec<RawStreamEvent> = serde_json::from_value(response["events"].clone())?;
        Ok(raw_events.into_iter().map(Into::into).collect())
    }

    /// Get stream statistics
    pub async fn stats(&self, room: &str) -> Result<StreamStats> {
        let payload = json!({"room": room});
        let response = self.client.send_command("stream.stats", payload).await?;
        Ok(serde_json::from_value(response)?)
    }

    /// List all stream rooms
    pub async fn list(&self) -> Result<Vec<String>> {
        let response = self.client.send_command("stream.list", json!({})).await?;
        Ok(serde_json::from_value(response["rooms"].clone())?)
    }

    /// Delete a stream room
    pub async fn delete_room(&self, room: &str) -> Result<()> {
        let payload = json!({"room": room});
        self.client.send_command("stream.delete", payload).await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::SynapConfig;

    #[test]
    fn test_stream_manager_creation() {
        let config = SynapConfig::new("http://localhost:15500");
        let client = SynapClient::new(config).unwrap();
        let stream = client.stream();

        assert!(std::mem::size_of_val(&stream) > 0);
    }

    #[test]
    fn test_stream_manager_clone() {
        let config = SynapConfig::new("http://localhost:15500");
        let client = SynapClient::new(config).unwrap();
        let stream1 = client.stream();
        let stream2 = stream1.clone();

        assert!(std::mem::size_of_val(&stream1) > 0);
        assert!(std::mem::size_of_val(&stream2) > 0);
    }
}
